tests/integration/test_kube_ovn.py

```diff
@@ -569,14 +595,57 @@ async def run_bird_curl_test(ops_test, unit, ip_to_curl):
 
 
 @pytest.mark.usefixtures("bird")
-async def test_bgp(ops_test, ips_to_curl):
+@pytest.mark.parametrize("scope", ["pod", "subnet"])
+async def test_bgp(ops_test, kubectl, kubectl_get, scope):
+    template_path = Path.cwd() / "tests/data/test-bgp.yaml"
+    template = template_path.read_text()
+    manifest = ops_test.tmp_path / "test-bgp.yaml"
+    manifest_data = template.format(
+        pod_annotations='annotations: {ovn.kubernetes.io/bgp: "true"}'
+        if scope == "pod"
+        else "",
+        subnet_annotations='annotations: {ovn.kubernetes.io/bgp: "true"}'
+        if scope == "subnet"
+        else "",
+    )
+    manifest.write_text(manifest_data)
+
+    async def cleanup():
+        await kubectl("delete", "--ignore-not-found", "-f", manifest)
+
+    await cleanup()
+
+    await kubectl("apply", "-f", manifest)
+    ips_to_curl = []
+    deadline = time.time() + 600
+
+    while time.time() < deadline:
+        pod = await kubectl_get("po", "-n", "test-bgp", "nginx")
+        pod_ip = pod.get("status", {}).get("podIP")
+        if pod_ip:
+            ips_to_curl.append(pod_ip)
+            break
+        log.info("Waiting for nginx pod IP")
+        await asyncio.sleep(1)
+
+    while time.time() < deadline:
+        svc = await kubectl_get("svc", "-n", "test-bgp", "nginx")
+        svc_ip = svc.get("spec", {}).get("clusterIP")
+        if svc_ip:
+            ips_to_curl.append(svc_ip)
+            break
+        log.info("Waiting for nginx svc IP")
+        await asyncio.sleep(1)
+
     log.info("Verifying the following IPs are reachable from bird units ...")
     log.info(ips_to_curl)
     bird_app = ops_test.model.applications["bird"]
     for unit in bird_app.units:
         for ip in ips_to_curl:
             assert await run_bird_curl_test(ops_test, unit, ip)
 
+    await cleanup()
+
```

Comment:
so far as i can tell -- this is the only `assert` in this test.  Assuming it fails to test -- the `cleanup()` below won't run


```suggestion
    try:
      bird_app = ops_test.model.applications["bird"]
      for unit in bird_app.units:
          for ip in ips_to_curl:
              assert await run_bird_curl_test(ops_test, unit, ip)
    finally:
      await cleanup()

```

Reply:
I've been back and forth on this line of thought for a long time. Should a failed test clean up after itself?

I think in an ideal world, on failure, the test would gather and output all information needed to determine why it failed, and then clean up after itself to leave the environment in a clean state.

In practice, I almost always find that test failures do not fully capture all of the information needed to determine the cause. Then, if the test cleans up after itself, the information is lost. It becomes impossible to determine the cause via manual inspection or via crashdumps that are captured later in the test suite.

So, my decision here to not clean up on failure is deliberate. If the test fails, I want the pod and subnet to stick around so that they can be inspected later.

Reply:
Great thought -- i can dig it!  👍🏻